### PR TITLE
#10442 Preferences title isn't reupdated after opening a screen optio…

### LINKED ISSUE #10442: [Bug] Preferences title isn't reupdated after opening a screen option
###### Reproduction Steps

1. Open preferences
2. Go to a preference which have its own page (e.g. Advanced > Advanced Statistics or Advanced > Custom sync server)
3. Return to the screen

https://user-images.githubusercontent.com/69634269/156920374-32cb434a-37ef-4054-ab26-cf7031681fb5.mp4

###### Expected Result

The preference title should reappear

###### Actual Result

The subpage title persists

Screenshot_20220306-080609_AnkiDroid

###### Debug info
```
AnkiDroid Version = 2.16alpha48

Android Version = 11

Manufacturer = samsung

Model = SM-N770F

Hardware = exynos9810

Webview User Agent = Mozilla/5.0 (Linux; Android 11; SM-N770F Build/RP1A.200720.012; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/99.0.4844.58 Mobile Safari/537.36

ACRA UUID = 692f00ed-cfa8-4d52-85d9-725a4f4e0fbb

Scheduler = std2

Crash Reports Enabled = true

DatabaseV2 Enabled = true
 ```

###### Research
*Enter an [x] character to confirm the points below:*

- [X] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [X] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [X] I have searched for similar existing issues here and on the user forum
- [X] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by BishyanKar**
1 bug fixed

## Purpose / Description
Preferences title isn't reupdated after navigating back from any of the inner fragments(like AdvancedStatisticsSettingsFragment), bug is fixed in this commit.

## Fixes
Fixes #10442

## Approach
Added onBackStackChangedListener to supportFragmentManager and based on the type of current fragment visible to user changed the title on its action bar.

## How Has This Been Tested?

Previous automated tests should be enough. Also manually tested by going back and forth between affected fragments.

## Learning (optional, can help others)
_Describe the research stage_

_Links to blog posts, patterns, libraries or addons used to solve this problem_

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

https://user-images.githubusercontent.com/54733471/157274683-6a0045b0-fc90-40ac-9499-170e5714d45a.mp4

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt [hunk 1]
```diff
@@ -93,6 +94,8 @@ class Preferences : AnkiActivity() {
     /** The collection path when Preferences was opened   */
     private var mOldCollectionPath: String? = null
 
+    private lateinit var mOnBackStackChangedListener: FragmentManager.OnBackStackChangedListener
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:97]: Does this need to be `lateinit`? Can the variable be initialized here as a `val`?

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt [hunk 2]
```diff
@@ -116,6 +119,27 @@ class Preferences : AnkiActivity() {
...
+
+    override fun onDestroy() {
+        super.onDestroy()
+        supportFragmentManager.removeOnBackStackChangedListener(mOnBackStackChangedListener)
+    }
+
+    private fun updateActionBarTitle(fragmentManager: FragmentManager, actionBar: ActionBar?) {
+        val fragment = fragmentManager.findFragmentById(R.id.settings_container)
+
+        when (fragment) {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:137]: You should be able to extract the assignment to outside the `when`:

```kotlin
actionBar?.title = when (fragment) { ...
```

**Review by david-allison — CHANGES_REQUESTED** [review]
Really high quality, I like this. There's one question and one nitpick, and after that, it should be good to go!

> Previous automated tests should be enough.

`Preferences` is really badly tested, but I trust this is correct, especially as it's been tested

I've kicked off CI to see if there's any issues.

Thanks so much!

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt [hunk 3]
```diff
@@ -116,6 +119,27 @@ class Preferences : AnkiActivity() {
...
+        supportFragmentManager.addOnBackStackChangedListener(mOnBackStackChangedListener)
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        supportFragmentManager.removeOnBackStackChangedListener(mOnBackStackChangedListener)
+    }
+
+    private fun updateActionBarTitle(fragmentManager: FragmentManager, actionBar: ActionBar?) {
+        val fragment = fragmentManager.findFragmentById(R.id.settings_container)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:135]: Might as well be explicit and return early if `actionBar` is null

Comment (BishyanKar) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:133]: Done

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt (hunk 2 above)
Comment (BishyanKar) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt]: Done

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt (hunk 1 above)
Comment (BishyanKar) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt]: Done

**Comment by BishyanKar** [comment]
Thank you for point these out, made the requested changes.

**Review by david-allison — APPROVED** [review]
Looks great! If CI is happy (and the changes are tested, I'm happy)

**Comment by mikehardy** [comment]
Nice! Thanks for the cleanup, much appreciated 🧹 🧹

**Review by mikehardy — APPROVED** [review]

**Comment by BishyanKar** [comment]
@david-allison please can you trigger the CI again? Last commit missed few updates from master so failed few checks.

**Comment by david-allison** [comment]
> @david-allison please can you trigger the CI again? Last commit missed few updates from main so failed few checks.

Done! It's a GitHub security feature for the first commit, you'll be able to do so yourself after this PR.

Just to note: we `rebase` onto `main`, rather than `merge`. No action needed with this PR as we'll squash-merge it, but it's worthwhile to consider in future.

We're working on documentation for rebasing, and we're more than open to discuss it on [Discord](https://discord.gg/qjzcRTx).

Thanks again for the PR, very clean.

**Pull request merged**